internal/batches/executor/execution_cache.go

```diff
@@ -37,14 +36,86 @@ func (key ExecutionCacheKey) Key() (string, error) {
 	// if an unrelated environment variable changes, that's fine. We're only
 	// interested in the ones that actually make it into the step container.
 	global := os.Environ()
-	envs := make([]map[string]string, len(key.Task.Steps))
-	for i, step := range key.Task.Steps {
+	envs := make([]map[string]string, len(steps))
+	for i, step := range steps {
 		env, err := step.Env.Resolve(global)
 		if err != nil {
-			return "", errors.Wrapf(err, "resolving environment for step %d", i)
+			return nil, errors.Wrapf(err, "resolving environment for step %d", i)
 		}
 		envs[i] = env
 	}
+	return envs, nil
+}
+
+func marshalHash(t *Task, envs []map[string]string) (string, error) {
+	raw, err := json.Marshal(struct {
+		*Task
+		Environments []map[string]string
+	}{
+		Task:         t,
+		Environments: envs,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	hash := sha256.Sum256(raw)
+	return base64.RawURLEncoding.EncodeToString(hash[:16]), nil
+}
+
+// StepsCachekey implements the CacheKeyer interface for a Task and a *subset*
+// of its Steps, up to and including the step with index StepIndex in
+// Task.Steps.
+type StepsCacheKey struct {
+	*Task
+	StepIndex int
+}
+
+// Key converts the key into a string form that can be used to uniquely identify
+// the cache key in a more concise form than the entire Task.
+func (key StepsCacheKey) Key() (string, error) {
+	// Setup a copy of the Task that only includes the Steps up to and
+	// including key.StepIndex.
+	taskCopy := &Task{
+		Repository:            key.Task.Repository,
+		Path:                  key.Task.Path,
+		OnlyFetchWorkspace:    key.Task.OnlyFetchWorkspace,
+		BatchChangeAttributes: key.Task.BatchChangeAttributes,
+		Template:              key.Task.Template,
+		TransformChanges:      key.Task.TransformChanges,
+		Archive:               key.Task.Archive,
+	}
+
+	taskCopy.Steps = key.Task.Steps[0 : key.StepIndex+1]
+
+	// Resolve environment only for the subset of Steps
+	envs, err := resolveStepsEnvironment(taskCopy.Steps)
+	if err != nil {
+		return "", err
+	}
+
+	hash, err := marshalHash(taskCopy, envs)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s-step-%d", hash, key.StepIndex), err
+}
+
+func (key StepsCacheKey) Slug() string { return key.Repository.Slug() }
```

Comment:
So, [when embedding structs in other structs in Go](https://eli.thegreenplace.net/2020/embedding-in-go-part-1-structs-in-structs/) you technically don't have to do `key.Task`, since `key.FooFooFoo()` would call `key.Task.FooFooFoo()`. 

I don't have a hard rule when to use one or the other, but here my thought process was: it's a 1-line method and it's only accessing a single method on a single field. Skipping the explicit `.Task` is fine - we're taking a shortcut without much danger. In the `Key()` method above, though, we have multiple "tasks": the `taskCopy` and the `key.Task` and we even change the latter to the be the former in a way. In that case I wanted to be explicit for which task is being accessed, so I used `key.Task`.